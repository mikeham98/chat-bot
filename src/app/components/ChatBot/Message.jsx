import React from 'react';
import PropTypes from 'prop-types';
import styles from '../../../themes/components/messages.scss';
import isPopulatedArray from "../../util/isPopulatedArray";
import Media from "./Media/Media";
import Options from "./Options/Options";
import {MediaObjectPropTypes, OptionsPropTypes} from "../../config/propTypes";

export default class Message extends React.PureComponent {
    constructor(props) {
        super(props);
        this.onClickOption = this.onClickOption.bind(this);
    }

    onClickOption(optionId) {
        const {id} = this.props;
        this.props.onClickOption(optionId, id)
    }

    showMedia() {
        return isPopulatedArray(this.props.media);
    }

    showOptions() {
        return isPopulatedArray(this.props.options);
    }

    render() {
        const {body, showDateTime, dateTime, media, options, onClickBody, selectedOption, color} = this.props;
        return (
            <div className={styles.messageContent}>
                <div className={styles.message} style={{backgroundColor: color}} onClick={onClickBody}>
                    <span className={styles.messageBody}>{body}</span>
                    {this.showOptions() && (
                        <Options
                            selectedOption={selectedOption}
                            options={options}
                            onClick={this.onClickOption}
                        />
                    )}
                </div>
                {this.showMedia() && (
                    <Media
                        media={media}
                    />
                )}
                {showDateTime && <span className={styles.messageCreatedAt}>{dateTime}</span>}
            </div>
        );
    }
}

Message.defaultProps = {
    color: ''
};

Message.propTypes = {
    onClickOption: PropTypes.func.isRequired,
    id: PropTypes.number,
    body: PropTypes.string,
    showDateTime: PropTypes.bool,
    dateTime: PropTypes.string,
    media: MediaObjectPropTypes,
    options: OptionsPropTypes,
    onClickBody: PropTypes.func.isRequired,
    selectedOption: PropTypes.number,
    color: PropTypes.string,
};