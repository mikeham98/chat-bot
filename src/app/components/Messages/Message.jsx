import React from 'react';
import styles from '../../../themes/messages.scss';

const Message = ({body, dateTime}) => {
    return (
        <div className={styles.message} data-tip={dateTime}>
            {body}
        </div>
    );
};

export default React.memo(Message);