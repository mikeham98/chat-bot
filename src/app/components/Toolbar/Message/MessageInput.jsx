import React from 'react';
import classnames from 'classnames';
import styles from '../../../../themes/toolbar.scss';

class MessageInput extends React.PureComponent {
    displayError() {
        const {error, touched} = this.props;
        return error && touched;
    };

    returnClassName() {
        const classNames = [styles.messageInputWrapper];
        if (this.displayError()) {
            classNames.push(styles.messageInputError);
        }
        return classnames(classNames);
    }

    render() {
        const {value, onChange, onBlur} = this.props;
        return (
            <input
                id={'message'}
                name={'message'}
                placeholder={'Type a message...'}
                type={'text'}
                value={value}
                onChange={onChange}
                onBlur={onBlur}
                className={this.returnClassName()}
            />
        );
    }
}

export default MessageInput;