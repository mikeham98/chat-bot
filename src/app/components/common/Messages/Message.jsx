import React from 'react';
import styles from '../../../../themes/messages.scss';

const Message = ({body}) => {
    return (
        <div className={styles.message}>
            {body}
        </div>
    );
};

export default React.memo(Message);