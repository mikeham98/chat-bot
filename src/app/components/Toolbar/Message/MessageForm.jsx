import React from 'react';
import {Formik} from 'formik';
import MessageInput from './MessageInput';
import IconButton from "../../common/Icon/Button";
import sendLogo from '../../../../../assets/icons/right-arrow-1.svg';
import styles from '../../../../themes/toolbar.scss';

export default class MessageForm extends React.Component {
    constructor(props) {
        super(props);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleSubmit({message}, {setSubmitting, resetForm}) {
        const callback = () => {
            setSubmitting(false);
            resetForm()
        };
        this.props.sendMessage(message, callback);
    }

    render() {
        return (
            <Formik
                initialValues={this.props.initialValues}
                onSubmit={this.handleSubmit}
                // validationSchema={Yup.object().shape({
                //     email: Yup.string()
                //         .email()
                //         .required('Required'),
                // })}
            >
                {formikProps => {
                    const {
                        values,
                        touched,
                        errors,
                        isSubmitting,
                        handleChange,
                        handleBlur,
                        handleSubmit,
                    } = formikProps;
                    return (
                        <form className={styles.messageFormWrapper} onSubmit={handleSubmit}>
                            <MessageInput
                                value={values.message}
                                onChange={handleChange}
                                onBlur={handleBlur}
                                touched={touched.message}
                                error={errors.message}
                            />
                            <IconButton
                                type='submit'
                                icon={sendLogo}
                                style={{margin: '0 10px'}}
                                color={values.message.length ? '#172B4D' : '#939aab'}
                                height='30px'
                                width='30px'
                                disabled={isSubmitting}
                            />
                        </form>
                    );
                }}
            </Formik>
        )
    }
}

MessageForm.defaultProps = {
    initialValues: {
        message: ''
    }
};