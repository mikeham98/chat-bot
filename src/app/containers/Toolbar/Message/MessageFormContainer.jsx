import React from 'react';
import {connect} from 'react-redux';
import MessageForm from '../../../components/Toolbar/Message/MessageForm';
import {sendMessage} from '../../../actions/toolbar/index.actions';

const mapDispatchToProps = (dispatch, ownProps) => {
    return {
        sendMessage: (message, callback) => {
            dispatch(sendMessage(ownProps.conversationId, message, callback))
        }
    };
};

export default connect(undefined, mapDispatchToProps)(MessageForm);