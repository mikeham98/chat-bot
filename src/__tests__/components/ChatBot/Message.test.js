import React from 'react';
import {shallow} from 'enzyme';
import Message from '../../../app/components/ChatBot/Message';

describe('Message', () => {

    let shallowedWrapper;
    let props = {};
    const wrapper = () => {
        if(!shallowedWrapper) {
            shallowedWrapper = shallow(<Message {...props}/>);
        }
        return shallowedWrapper;
    };

    beforeEach(() => {
        props = {
            body: 'Hello world',
            dateTime: '10 minutes ago',
            showDateTime: false
        };
        shallowedWrapper = undefined;
    });

    describe('render', () => {
        it('should return a message body', () => {
            expect(wrapper()).toMatchSnapshot();
        });
        it('should return a message body along with showDateTime', () => {
            wrapper().setProps({
                showDateTime: true
            });
            expect(wrapper()).toMatchSnapshot();
        });
    });
});