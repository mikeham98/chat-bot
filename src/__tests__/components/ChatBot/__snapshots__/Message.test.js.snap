// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message render should return a message body with media and options 1`] = `
<div
  className="messageContent"
>
  <div
    className="message"
    onClick={[MockFunction]}
    style={
      Object {
        "backgroundColor": "",
      }
    }
  >
    <span
      className="messageBody"
    >
      Hello world
    </span>
    <Options
      onClick={[Function]}
      options={
        Array [
          1,
          2,
          3,
          4,
          5,
        ]
      }
      selectedOption={1}
    />
  </div>
  <Media
    media={
      Array [
        1,
        2,
      ]
    }
  />
</div>
`;

exports[`Message render should return a message body with media and options along with showDateTime 1`] = `
<div
  className="messageContent"
>
  <div
    className="message messageFocused"
    onClick={[MockFunction]}
    style={
      Object {
        "backgroundColor": "",
      }
    }
  >
    <span
      className="messageBody"
    >
      Hello world
    </span>
    <Options
      onClick={[Function]}
      options={
        Array [
          1,
          2,
          3,
          4,
          5,
        ]
      }
      selectedOption={1}
    />
  </div>
  <Media
    media={
      Array [
        1,
        2,
      ]
    }
  />
  <span
    className="messageCreatedAt"
  >
    10 minutes ago
  </span>
</div>
`;

exports[`Message render should return a message body without media 1`] = `
<div
  className="messageContent"
>
  <div
    className="message"
    onClick={[MockFunction]}
    style={
      Object {
        "backgroundColor": "",
      }
    }
  >
    <span
      className="messageBody"
    >
      Hello world
    </span>
    <Options
      onClick={[Function]}
      options={
        Array [
          1,
          2,
          3,
          4,
          5,
        ]
      }
      selectedOption={1}
    />
  </div>
</div>
`;

exports[`Message render should return a message body without options 1`] = `
<div
  className="messageContent"
>
  <div
    className="message"
    onClick={[MockFunction]}
    style={
      Object {
        "backgroundColor": "",
      }
    }
  >
    <span
      className="messageBody"
    >
      Hello world
    </span>
  </div>
  <Media
    media={
      Array [
        1,
        2,
      ]
    }
  />
</div>
`;
